taichi/ir/state_machine.h

```diff
@@ -0,0 +1,82 @@
+#pragma once
+
+#include "taichi/ir/ir.h"
+#include <unordered_set>
+
+TLANG_NAMESPACE_BEGIN
+
+// State machine for AllocaStmt/GlobalTemporaryStmt/GlobalPtrStmt.
+class StateMachine {
+ private:
+  Stmt *var;
+  static std::unique_ptr<std::unordered_set<AtomicOpStmt *>> used_atomics;
+
+  bool same_data(Stmt *store_stmt1, Stmt *store_stmt2);
+
+ public:
+  // If neither stored nor loaded (nor used as operands in masks/loop_vars),
+  // we can safely delete this variable if it's an alloca or a global temp.
+  enum State {
+    never,
+    maybe,
+    definitely
+  };
+  State stored;  // Is this variable ever stored (or atomic-operated)?
+  State stored_in_this_if_or_loop;
+  State loaded;  // Is this variable ever loaded (or atomic-operated)?
+  State loaded_in_this_if_or_loop;
+
+  static State merge(const State &a, const State &b) {
+    if (a == definitely || b == definitely)
+      return definitely;
```

Comment:
Should this be `a == definitely && b == definitely`?

Reply:
You're right. `stored = merge(true_branch.stored, false_branch.stored);` and `merge(true_branch.stored_in_this_if_or_loop, false_branch.stored_in_this_if_or_loop)` should be `a == definitely && b == definitely`, but `stored_in_this_if_or_loop = merge(stored_in_this_if_or_loop, merge_result` should be `a == definitely || b == definitely`...

Reply:
I feel like `stored_in_this_if_or_loop` should not be a `State`, but two `State`s. This might help clarify the logic here...

Reply:
I named it like this because we can't eliminate `a = 1;` in these cases:
```C++
a = 1;
if (...)
  a = 2;
```
```C++
a = 1;
for (...)
  a = 2;
```
But we *can* eliminate `a = 1` in this case:
```C++
a = 1;
{
  a = 2;
}
```

Reply:
It's just `if` and `for` are "maybe" executed.

Reply:
I see. Do you have a moment to briefly chat about these on zoom?

Reply:
Sure. (I've just pushed a commit about these)